### PR TITLE
LNS

### PR DESCRIPTION
**Pull request by majestrate**
relay and client side LNS.

### CODE REVIEW

File: llarp/service/tag_lookup_job.hpp [hunk 1]
```diff
@@ -1,75 +1,4 @@
 #ifndef LLARP_SERVICE_TAG_LOOKUP_JOB_HPP
 #define LLARP_SERVICE_TAG_LOOKUP_JOB_HPP
```
Comment (tewinget) [inline llarp/service/tag_lookup_job.hpp:2]: delete it harder!

File: llarp/service/endpoint_types.hpp [hunk 2]
```diff
@@ -54,6 +54,8 @@ namespace llarp
 
     using PathEnsureHook = std::function<void(Address, OutboundContext*)>;
 
+    using LNSNameCache = std::unordered_map<std::string, std::pair<Address, llarp_time_t>>;
```
Comment (tewinget) [inline llarp/service/endpoint_types.hpp:57]: Should probably either make decayed hashset a templated type with an arbitrary container, or make a decayed map class.  Likely other things will come along (or are already implemented) that would benefit from such a class.

File: llarp/rpc/lokid_rpc_client.cpp [hunk 3]
```diff
@@ -215,6 +215,36 @@ namespace llarp
...
+              try
+              {
+                const auto j = nlohmann::json::parse(data[1]);
+                const auto itr = j.find("encrypted_value");
+                if (itr != j.end())
+                {
+                  result = lokimq::from_hex(itr->get<std::string>());
+                }
+              }
+              catch (...)
```
Comment (tewinget) [inline llarp/rpc/lokid_rpc_client.cpp:239]: Should have a comment with a `catch (...) do nothing` block, to save others and future-you from figuring out what you're ignoring and why.

File: llarp/handlers/tun.cpp [hunk 4]
```diff
@@ -468,6 +469,21 @@ namespace llarp
                 addr.as_array(), std::make_shared<dns::Message>(msg), isV6);
           }
         }
+        else if (ends_with(qname, ".loki"))
+        {
+          return LookupNameAsync(
```
Comment (tewinget) [inline llarp/handlers/tun.cpp:474]: maybe length check here and NX if the name is too long?

Comment (jagerman) [reply llarp/handlers/tun.cpp:474]: Validity rules for registration:
- not `localhost.loki`, `loki.loki`, or `snode.loki`
- if it contains no dash then max 32 characters long (assuming a leading subdomain has been stripped)
These are from general DNS requirements, and also enforced in registrations:
- Must be all [A-Za-z0-9-].  (A-Z will be lower-cased by the RPC call).
- cannot start or end with a `-`
- max 63 characters long if it does contain a dash
- cannot contain `--` in the third and fourth characters *unless* it starts with `xn--`

File: llarp/dht/messages/findname.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    if (key == "T")
+    {
+      return bencode_read_integer(val, &TxID);
+    }
+    return bencode_discard(val);
+  }
+
+  bool
+  FindNameMessage::HandleMessage(struct llarp_dht_context* dht, std::vector<Ptr_t>& replies) const
```
Comment (tewinget) [inline llarp/dht/messages/findname.cpp:42]: a bit of whitespace between sections in here would help with readability imo

File: llarp/dht/messages/gotname.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#include <dht/messages/gotname.hpp>
+#include <lokimq/bt_serialize.h>
+
+namespace llarp::dht
+{
+  constexpr size_t NameSizeLimit = 512;
```
Comment (tewinget) [inline llarp/dht/messages/gotname.cpp:6]: as per conversation about this, seems checking against this limit doesn't make sense here.

File: llarp/service/tag_lookup_job.hpp (hunk 1 above)
Comment (majestrate) [reply llarp/service/tag_lookup_job.hpp]: done

File: llarp/handlers/tun.cpp (hunk 4 above)
Comment (majestrate) [reply llarp/handlers/tun.cpp:475]: added this check

File: llarp/service/endpoint_types.hpp (hunk 2 above)
Comment (majestrate) [reply llarp/service/endpoint_types.hpp:57]: refactored

File: llarp/dht/messages/gotname.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      : IMessage(from), result(std::move(data)), TxID(txid)
+  {
+    if (result.ciphertext.size() > NameSizeLimit)
+      throw std::invalid_argument("name data too big");
+  }
+
+  bool
+  GotNameMessage::BEncode(llarp_buffer_t* buf) const
+  {
+    const std::string nonce((const char*)result.nonce.data(), result.nonce.size());
```
Comment (tewinget) [inline llarp/dht/messages/gotname.cpp:18]: Encoding this as a string rather than an integer won't run into endian-ness issues?

Comment (tewinget) [reply llarp/dht/messages/gotname.cpp:18]: Nevermind, I'm dumb.

**Review by tewinget — APPROVED** [review]
LGTM

**Pull request merged**